packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6278,50 +6281,197 @@ abstract class RenderObjectElement extends Element {
+    Element? ancestor = _parent;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> debugAncestorTypes = <Type>{};
+    final Set<Type> debugParentDataTypes = <Type>{};
+    final List<Type> debugAncestorCulprits = <Type>[];
+    final List<Type> debugParentDataCulprits = <Type>[];
+
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidget types.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. Each contributing ParentDataWidget must contribute to a unique
+    //    ParentData type, less ParentData be overwritten.
+    //    For example, there cannot be two ParentDataWidgets that both write
+    //    ParentData of type KeepAliveParentDataMixin, if the first check was
+    //    subverted by a subclassing of the KeepAlive ParentDataWidget.
+    // 3. The ParentData itself must be compatible with all ParentDataWidgets
+    //    writing to it.
+    //    For example, TwoDimensionalViewportParentData uses the
+    //    KeepAliveParentDataMixin, so it could be compatible with both
+    //    KeepAlive, and another ParentDataWidget with ParentData type
+    //    TwoDimensionalViewportParentData or a subclass thereof.
+    // The first and second cases are verified here. The third is verified in
+    // debugIsValidRenderObject.
+
+    while (ancestor != null && ancestor is! RenderObjectElement) {
+      if (ancestor is ParentDataElement<ParentData>) {
+        assert(() {
+          if (!debugAncestorTypes.add(ancestor.runtimeType)) {
+            if (!debugAncestorCulprits.contains(ancestor.runtimeType)) {
+              // Add the first one we had put in the Set of Types.
+              debugAncestorCulprits.add(ancestor.runtimeType);
+            }
+            debugAncestorCulprits.add(ancestor.runtimeType);
+          }
+
+          if (!debugParentDataTypes.add((ancestor! as ParentDataElement<ParentData>).parentDataType)) {
+            if (!debugParentDataCulprits.contains((ancestor as ParentDataElement<ParentData>).parentDataType)) {
+              // Add the first one we had put in the Set of Types.
+              debugParentDataCulprits.add(ancestor.parentDataType);
+            }
+            debugParentDataCulprits.add(ancestor.parentDataType);
+          }
+          return true;
+        }());
+        result.add(ancestor);
+      }
+      ancestor = ancestor._parent;
+    }
+    assert(() {
+      if (result.isEmpty || ancestor == null) {
+        return true;
+      }
+      // Validate point 1 from above.
+      _debugCheckCompetingAncestors(
+        result,
+        debugAncestorTypes,
+        debugAncestorCulprits,
+      );
+      // Validate point 2 from above.
+      _debugCheckCompetingParentData(
```

Comment:
I can't seem to think of a case that fails point 1, but doesn't fail point 2. Unless I am missing something, I would only check for point 2. That would also solve the double exception problem I was "complaining" about below. We would have to enrich the point 2 exception with some of the information from the point 1 exception, though.

Reply:
Yeah I realized this while I was writing the tests and wondered what you might think. I agree. The subclasses and custom Positioned test cases showed that even when the Type of ParentDataWidget is not the same, it is always a subclass of ParentDataWidget<SomeParentData>, so the SomeParentData is where this really matters.